docs/design/2021-09-13-parser-as-submodule-of-tidb.md

```diff
@@ -0,0 +1,89 @@
+
+Personally, that is a weird argument for me. Indeed, pingcap/tidb does not store the diff of `parser.go` anymore, but pingcap/parser does. How is moving the inflation from one repo to another better?
+
+Another argument is about the messy dependency. The separation does improve the case.
+
+### What is wrong with a separate repository?
+
+In fact, there is no big problem. However the development becomes inconvenient. You could argue that we have workarounds, but why not make the development easier?
+
+1. We always need to ensure the version of two go modules. Cherry-picking becomes troublesome since the import version must be a specific branch of pingcap/parser.
+
+2. It is a mysql parser, but is closely related to TiDB(basically everything except ast/mysql packages). You must code TiDB logic into the "standalone mysql parser" first before your real implementation, which means at least two PRs, even for small changes.
+
+3. Before the PR of parser merged, the PR of TiDB must use module directive to redirect a specific version of parser to fix CI.
+
+Problems above have defeated the previous argument `only go.mod need to change accordingly`. We only separated code into two repositories, actual work did not decrease. However, our development experience does getting worse.
+
+## Detailed Plan
+
+Obviously, we could move back to the old good way. But it must be a go sub-module, because:
+
+1. To keep compatibility. Parser has been imported by other tools and other users.
+2. To keep the clean dependency. That is why it is separated. We don't want to make things worse again.
+
+The detailed plan:
+
+1. Create parser directory, a sub-module, in pingcap/tidb, which is synchronized with pingcap/parser, with `replace github.com/pingcap/parser => ./parser`.
+2. Stop sending new PRs and issues to pingcap/parser, new PRs should be directly sent to pingcap/tidb. And merge recent PRs in pingcap/parser as much as possible. Changes will be synchronized from pingcap/parser to pingcap/tidb(manually), of course.
+3. Announce the deprecation of pingcap/parser to possible users. Clean up old PRs, reopen it in pingcap/tidb, close it, or just forget about it. As for issues, by @kennytm, we could directly transfer all issues to pingcap/tidb according to the [document](https://docs.github.com/en/issues/tracking-your-work-with-issues/transferring-an-issue-to-another-repository).
+4. We can create wrapper packages for pingcap/parser, which re-exports things from `pingcap/tidb/parser`. I mean something like `import ( . github.com/pingcap/tidb/parser/xxx)`, create a dummy function will eliminate the error of `not used import`. All code of pingcap/parser will be removed in this step. This will delevery new updates without the need of migrating importing paths.
+5. After another one or two dev cycles, we could archive pingcap/parser. It mainly depends on users of pingcap/parser.
+
```

Comment:
You should also mention how to migrate the test infra.

Reply:
OK I see it below, but please consider it in the implementation plan. It is ended by migrated, not only answered the question.

Reply:
Added in step 3.